File: 564. Find the Closest Palindrome.py
class Solution:
    def nearestPalindromic(self, S):
        N = len(S)
        Val = lambda x: abs(int(S) - int(x)) 
        P = int(S[:(N + 1) >> 1])
        Dp = [i + (i[:len(i)-(N&1)])[::-1] for i in map(str, range(P - 1, P + 2))] + ['1' + '0'*(N-1) + '1']
        ans = '9' * (N - 1) or '0'
        for i in Dp:
            if not (i.startswith('00') or i == S) and Val(i) < Val(ans):
                ans = i
        return ans

File: 564. Find the Closest Palindrome.java
class Solution {
    public String nearestPalindromic(String n) {
        int size = n.length(); 
        int mid = size/2;

        long firstHalf = Long.parseLong(n.substring(0, size % 2 == 0 ? mid : mid + 1));
       
        List<Long> possibleResults = new ArrayList<>();
        possibleResults.add(halfToPalindrome(firstHalf, size % 2 == 0));
        possibleResults.add(halfToPalindrome(firstHalf + 1, size % 2 == 0));
        possibleResults.add(halfToPalindrome(firstHalf - 1, size % 2 == 0));
        possibleResults.add((long) Math.pow(10, size - 1) - 1);  // Edge case like 999...
        possibleResults.add((long) Math.pow(10, size) + 1);      // Edge case like 100...001

         long diff = Long.MAX_VALUE;
        long result = 0;
        long original = Long.parseLong(n);
        for(long num:possibleResults)
        {
            if(num==original)
            continue;
            if(Math.abs(num-original)<diff)
            {
                diff = Math.abs(num-original);
                result = num;
            }
            else if(Math.abs(num-original)==diff)
            {
                result = Math.min(result,num);
            }
        }
        return String.valueOf(result);
    }
    long halfToPalindrome(long firstHalf,boolean bool)
    {
        long result = firstHalf;
        if(!bool)
        {
            firstHalf = firstHalf/10;
        }
         while(firstHalf>0)
         {
            int digit = (int)(firstHalf%10);
            result = result*10+digit;
            firstHalf = firstHalf/10;
         }
         return result;
    }
}
